### PR TITLE
Add smartify Liquid filter for SmartyPants

### PR DESCRIPTION
**Pull request by pathawks**
Is there a better place to put the `Kramdown::Parser::SmartyPants` class?

Am I on the right track with this?

### CODE REVIEW

**Comment by pathawks** [comment]
**Prior Art:**
- [**jekyll-smartify**](https://github.com/pathawks/jekyll-smartify)
- #3566
- #2766
- #69

**Comment by envygeeks** [comment]
This is a tough spot, I don't know how @jekyll/core likes file organization in that aspect.  For my own libs and apps, I would put it in `lib/kramdown/parser/smarty_pants` even if I had `lib/jekyll` but not all authors are strict in that way in that the class must point to the file or I'm doing it wrong.

**Comment by parkr** [comment]
How does the SmartyPants converter hook into `Kramdown::Parser::SmartyPants`?

**Comment by parkr** [comment]
Hm. Ok. Can it point to a class in our namespace? I'm generally weary of writing something in someone else's namespace.

**Comment by pathawks** [comment]
> Can it point to a class in our namespace?

Yes, good call. How does this look?

The failing tests freaked me out since they are related to Kramdown, but they are the same tests that [are failing on **master**, so...](https://travis-ci.org/jekyll/jekyll/jobs/100995381)

**Comment by pathawks** [comment]
There is [one test](https://travis-ci.org/jekyll/jekyll/jobs/100995443#L296) that fails on Ruby 2.0 and nothing else:

```text
Error:
TestFilters#test_: filters should smartify with simple string. :
NameError: wrong constant name Jekyll::Converters::SmartyPants::KramdownParser
    /home/travis/build/jekyll/jekyll/vendor/bundle/ruby/2.0.0/gems/kramdown-1.9.0/lib/kramdown/document.rb:106:in `const_defined?'
    /home/travis/build/jekyll/jekyll/vendor/bundle/ruby/2.0.0/gems/kramdown-1.9.0/lib/kramdown/document.rb:106:in `initialize'
    /home/travis/build/jekyll/jekyll/lib/jekyll/converters/smartypants.rb:23:in `new'
    /home/travis/build/jekyll/jekyll/lib/jekyll/converters/smartypants.rb:23:in `convert'
    /home/travis/build/jekyll/jekyll/lib/jekyll/filters.rb:26:in `smartify'
    /home/travis/build/jekyll/jekyll/test/test_filters.rb:35:in `block (2 levels) in <class:TestFilters>'
    /home/travis/build/jekyll/jekyll/vendor/bundle/ruby/2.0.0/gems/shoulda-context-1.2.1/lib/shoulda/context/context.rb:413:in `instance_exec'
    /home/travis/build/jekyll/jekyll/vendor/bundle/ruby/2.0.0/gems/shoulda-context-1.2.1/lib/shoulda/context/context.rb:413:in `block in create_test_from_should_hash'
```

Any idea what could be tripping up Ruby 2.0?

**Comment by pathawks** [comment]
> Can it point to a class in our namespace? I'm generally weary of writing something in someone else's namespace.

According to [Kramdown](https://github.com/gettalong/kramdown/blob/master/lib/kramdown/document.rb#L95):

> It has to be the name of a class in the Kramdown::Parser module.

It seems to work *anyway*, but maybe this is the cause of the Ruby 2.0 fail?

**Comment by parkr** [comment]
> The special options key :input can be used to select the parser that should parse the source. It has to be the name of a class in the Kramdown::Parser module.

Ah, very interesting. So I guess we do. If it's that small, I don't mind it being in the smartypants converter.

File: lib/jekyll.rb [hunk 1]
```diff
@@ -172,6 +172,7 @@ def sanitized_path(base_directory, questionable_path)
 require_all 'jekyll/commands'
 require_all 'jekyll/converters'
 require_all 'jekyll/converters/markdown'
+require_all 'jekyll/converters/smartypants'
```
Comment (parkr) [inline lib/jekyll.rb]: this should be covered by line 173 :)

Comment (pathawks) [inline lib/jekyll.rb]: Line 173 included `jekyll/converters/smartypants.rb` but not `jekyll/converters/smartypants/kramdown_parser.rb`

Comment (envygeeks) [inline lib/jekyll.rb]: I think both those requires should be put into `jekyll/converters` tbh, they are subs of it.

Comment (parkr) [inline lib/jekyll.rb]: One is a sub of `Kramdown::Parser::Markdown`, another of `Jekyll::Converter`?

**Comment by pathawks** [comment]
> If it's that small, I don't mind it being in the smartypants converter.

In the same file?

**Comment by parkr** [comment]
> In the same file?

Yeah. It's like 5 lines, right?

File: lib/jekyll.rb (hunk 1 above)
Comment (envygeeks) [inline lib/jekyll.rb]: @parkr wat? I'm talking about the markdown and smartypants requires, both those belong in jekyll/coverters.rb.

**Comment by pathawks** [comment]
I have added several tests, and am now allowing HTML to passthrough. There are separate filters that will escape/strip HTML, if that is desired.

**Comment by pathawks** [comment]
Just for reference, I brought up the idea of adding this parser to kramdown proper.

gettalong/kramdown#299

**Comment by parkr** [comment]
@jekyllbot: merge +minor

**Comment by parkr** [comment]
Most useful robot EVER.

**Comment by ndarville** [comment]
Would *definitely* love an alternative to `{{ page.title | markdownify | strip_html` }}. :)

**Comment by jhabdas** [comment]
Sounds good to me. Need to grok our discussion in https://github.com/gettalong/kramdown/pull/299 a little more before I do. Thanks for staying on top of these issues and all of your helpful feedback, @pathawks.

**Pull request merged**